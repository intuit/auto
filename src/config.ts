/* tslint:disable:no-inferrable-types */
export enum Label {
  major = 'major',
  minor = 'minor',
  patch = 'patch',
  release = 'release',
  prerelease = 'prerelease',
  skipRelease = 'skipRelease'
}

export class Labels {
  public [Label.major]: string = 'major';
  public [Label.minor]: string = 'minor';
  public [Label.patch]: string = 'patch';
  public [Label.skipRelease]: string = 'skip-release';
  /**
   * @deprecated
   */
  public 'skip-release': string = 'skip-release';
  public [Label.release]: string = 'release';
  public [Label.prerelease]: string = 'prerelease';
}

export class ChangelogTitles {
  public major = '💥  Breaking Change';
  public minor = '🚀  Enhancement';
  public patch = '🐛  Bug Fix';
  public internal = '🏠  Internal';
  public documentation = '📝  Documentation';
  [label: string]: string;
}

export class Config {
  public noVersionPrefix?: boolean;
  public onlyPublishWithReleaseLabel?: boolean;
  public name?: string;
  public email?: string;
  public owner?: string;
  public repo?: string;
  public labels?: Labels = new Labels();
  public changelogTitles?: ChangelogTitles = new ChangelogTitles();
  public skipReleaseLabels?: string[] = [];
  public slack?: string;
  public jira?: string;
  public githubApi?: string;
}
