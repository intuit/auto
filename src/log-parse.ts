import { ICommit } from 'gitlog';
import { AsyncSeriesBailHook } from 'tapable';
import { URL } from 'url';
import join from 'url-join';
import { VersionLabel } from './github-release';
import { ILogger } from './utils/logger';
import { makeLogHooks } from './utils/make-hooks';

export interface ICommitAuthor {
  name?: string;
  email?: string;
  username?: string;
}

export interface IGenerateReleaseNotesOptions {
  owner: string;
  repo: string;
  baseUrl: string;
  jira?: string;
  changelogTitles: { [label: string]: string };
  versionLabels: Map<VersionLabel, string>;
}

export interface IPullRequest {
  number: number;
  base?: string;
}

export type IExtendedCommit = ICommit & {
  authors: ICommitAuthor[];
  pullRequest?: IPullRequest;
  jira?: {
    number: string[];
  };
  labels: string[];
  packages?: string[];
};

type CommitParseFunction = (
  commit: IExtendedCommit
) => IExtendedCommit | undefined;

export interface ILogParseHooks {
  renderChangelogLine: AsyncSeriesBailHook<
    [IExtendedCommit[], (commit: IExtendedCommit) => Promise<string>],
    string[] | void
  >;
  renderChangelogTitle: AsyncSeriesBailHook<
    [string, { [label: string]: string }],
    string | void
  >;
  renderChangelogAuthor: AsyncSeriesBailHook<
    [ICommitAuthor, IExtendedCommit, IGenerateReleaseNotesOptions],
    string | void
  >;
  renderChangelogAuthorLine: AsyncSeriesBailHook<
    [ICommitAuthor, string],
    string | void
  >;
}

export function filterServiceAccounts(
  commit: IExtendedCommit
): IExtendedCommit | undefined {
  const SERVICE_ACCOUNTS = ['pdbf'];

  if (commit.authorName && SERVICE_ACCOUNTS.includes(commit.authorName)) {
    return;
  }

  return commit;
}

export function parsePR(commit: IExtendedCommit): IExtendedCommit {
  const merge = /Merge pull request #(\d+) from (.+)\n([\S\s]+)/;
  const prMatch = commit.subject.match(merge);

  if (!prMatch) {
    return commit;
  }

  return {
    ...commit,
    pullRequest: {
      number: Number(prMatch[1]),
      base: prMatch[2]
    },
    subject: prMatch[3].trim()
  };
}

export function parseSquashPR(
  commit: IExtendedCommit
): IExtendedCommit | undefined {
  const firstLine = commit.subject.split('\n')[0];
  const squashMerge = /\(#(\d+)\)$/;

  const squashMergeMatch = firstLine.match(squashMerge);

  if (!squashMergeMatch) {
    return commit;
  }

  return {
    ...commit,
    pullRequest: {
      number: Number(squashMergeMatch[1])
    },
    subject: firstLine
      .substr(0, firstLine.length - squashMergeMatch[0].length)
      .trim()
  };
}

export function parseJira(commit: IExtendedCommit): IExtendedCommit {
  // Support 'JIRA-XXX:' and '[JIRA-XXX]' and '[JIRA-XXX] - '
  const jira = /^\[?([\w]{3,}-\d+)\]?:?\s?[-\s]*([\S ]+)?/;
  const matches = [];

  let currentMatch = commit.subject.match(jira);

  while (currentMatch) {
    matches.push(currentMatch);
    currentMatch = currentMatch[2].match(jira);
  }

  if (!matches.length) {
    return commit;
  }

  return {
    ...commit,
    jira: {
      number: matches.map(match => match[1])
    },
    subject: matches[matches.length - 1][2].trim()
  };
}

function normalizeCommit(commit: ICommit): IExtendedCommit | undefined {
  const parsers: CommitParseFunction[] = [
    filterServiceAccounts,
    parsePR,
    parseSquashPR,
    parseJira
  ];

  const extended: IExtendedCommit = {
    labels: [],
    ...commit,
    authors: [{ name: commit.authorName, email: commit.authorEmail }]
  };

  return parsers.reduce(
    (prev: IExtendedCommit | undefined, parser) =>
      prev === undefined ? undefined : parser(prev),
    extended
  );
}

export function normalizeCommits(commits: ICommit[]): IExtendedCommit[] {
  return commits.map(normalizeCommit).filter(Boolean) as IExtendedCommit[];
}

const filterLabel = (commits: IExtendedCommit[], label: string) =>
  commits.filter(commit => commit.labels.includes(label));

export default class LogParse {
  public readonly hooks: ILogParseHooks;

  private readonly logger: ILogger;
  private readonly options: IGenerateReleaseNotesOptions;

  constructor(logger: ILogger, options: IGenerateReleaseNotesOptions) {
    this.logger = logger;
    this.options = options;
    this.hooks = makeLogHooks();
    this.options.changelogTitles.pushToMaster = '⚠️  Pushed to master';
  }

  public loadDefaultHooks() {
    this.hooks.renderChangelogAuthor.tap('Default', (author, commit) =>
      this.createUserLink(author, commit)
    );
    this.hooks.renderChangelogAuthorLine.tap('Default', (author, user) => {
      const authorString =
        author.name && user ? `${author.name} (${user})` : user;
      return authorString ? `- ${authorString}` : undefined;
    });
    this.hooks.renderChangelogLine.tapPromise(
      'Default',
      async (currCommits, renderLine) =>
        Promise.all(currCommits.map(async commit => renderLine(commit)))
    );
    this.hooks.renderChangelogTitle.tap(
      'Default',
      (label, changelogTitles) => `#### ${changelogTitles[label]}\n`
    );
  }

  // Every Commit will either be a PR, jira story, or push to master (patch)
  public async generateReleaseNotes(
    commits: IExtendedCommit[]
  ): Promise<string> {
    if (commits.length === 0) {
      return '';
    }

    this.logger.verbose.info('Generating release notes for:\n', commits);

    const split = this.splitCommits(commits);

    this.logger.verbose.info('Split commits into groups');
    this.logger.veryVerbose.info('\n', split);

    const sections: string[] = [];

    await this.createLabelSection(split, sections);

    this.logger.verbose.info('Added groups to changelog');

    await this.createAuthorSection(split, sections);

    this.logger.verbose.info('Added authors to changelog');

    const result = sections.join('\n\n');

    this.logger.verbose.info('Successfully generated release notes.');

    return result;
  }

  public createUserLink(author: ICommitAuthor, commit: IExtendedCommit) {
    const githubUrl = new URL(this.options.baseUrl).origin;

    if (author.username === 'invalid-email-address') {
      return;
    }

    return author.username
      ? `[@${author.username}](${join(githubUrl, author.username)})`
      : author.email || commit.authorEmail;
  }

  /**
   * Split commits into changelogTitle sections.
   */
  private splitCommits(
    commits: IExtendedCommit[]
  ): { [key: string]: IExtendedCommit[] } {
    let currentCommits = [...commits];

    commits
      .filter(commit => commit.labels.length === 0)
      .map(commit => commit.labels.push('patch'));

    return Object.assign(
      {},
      ...Object.keys(this.options.changelogTitles).map(label => {
        const matchedCommits = filterLabel(
          currentCommits,
          this.options.versionLabels.get(label as VersionLabel) || label
        );

        if (matchedCommits.length === 0) {
          return {};
        }

        currentCommits = currentCommits.filter(
          commit => !matchedCommits.includes(commit)
        );

        return {
          [label]: matchedCommits
        };
      })
    );
  }

  private async createUserLinkList(commit: IExtendedCommit) {
    const result = new Set<string>();

    await Promise.all(
      commit.authors.map(async author => {
        const link = await this.hooks.renderChangelogAuthor.promise(
          author,
          commit,
          this.options
        );

        if (link) {
          result.add(link);
        }
      })
    );

    return [...result].join(' ');
  }

  private async generateCommitNote(commit: IExtendedCommit) {
    let jira = '';
    let pr = '';

    if (commit.jira && this.options.jira) {
      const link = join(this.options.jira, ...commit.jira.number);
      jira = `[${commit.jira.number}](${link}): `;
    }

    if (commit.pullRequest) {
      const prLink = join(
        this.options.baseUrl,
        'pull',
        commit.pullRequest.number.toString()
      );
      pr = `[#${commit.pullRequest.number}](${prLink})`;
    }

    const user = await this.createUserLinkList(commit);

    return `- ${jira}${commit.subject} ${pr}${user ? ` (${user})` : ''}`;
  }

  private async createAuthorSection(
    split: { [key: string]: IExtendedCommit[] },
    sections: string[]
  ) {
    const authors = new Set<string>();
    const commits = Object.values(split).reduce(
      (
        labeledCommits: IExtendedCommit[],
        sectionCommits: IExtendedCommit[]
      ) => [...labeledCommits, ...sectionCommits],
      []
    );

    await Promise.all(
      commits.map(async commit => {
        commit.authors.map(async author => {
          if (author.username === 'invalid-email-address') {
            return;
          }

          const user = await this.hooks.renderChangelogAuthor.promise(
            author,
            commit,
            this.options
          );
          const authorEntry = await this.hooks.renderChangelogAuthorLine.promise(
            author,
            user as string
          );

          if (authorEntry && !authors.has(authorEntry)) {
            authors.add(authorEntry);
          }
        });
      })
    );

    if (authors.size > 0) {
      let authorSection = `#### Authors: ${authors.size}\n\n`;
      authorSection += [...authors].join('\n');
      sections.push(authorSection);
    }
  }

  private async createLabelSection(
    split: { [key: string]: IExtendedCommit[] },
    sections: string[]
  ) {
    await Promise.all(
      Object.entries(split).map(async ([label, labelCommits]) => {
        const title = await this.hooks.renderChangelogTitle.promise(
          label,
          this.options.changelogTitles
        );
        const lines = await this.hooks.renderChangelogLine.promise(
          labelCommits,
          async commit => this.generateCommitNote(commit)
        );

        sections.push([title, ...lines].join('\n'));
      })
    );
  }
}
