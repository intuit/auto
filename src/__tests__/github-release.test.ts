import { Labels } from '../config';
import GitHubRelease from '../github-release';
import { normalizeCommits } from '../log-parse';
import SEMVER from '../semver';
import { dummyLog } from '../utils/logger';
import makeCommitFromMsg from './make-commit-from-msg';

const constructor = jest.fn();
const getGitLog = jest.fn();
const publish = jest.fn();
const getLabels = jest.fn();
const getPullRequest = jest.fn();
const getLatestRelease = jest.fn();
const getSha = jest.fn();
const createStatus = jest.fn();
const getProject = jest.fn();
const createComment = jest.fn();
const changedPackages = jest.fn();
const getCommitsForPR = jest.fn();
const getUserByEmail = jest.fn();
const getUserByUsername = jest.fn();
const getProjectLabels = jest.fn();
const createLabel = jest.fn();
const getPullRequests = jest.fn();

getProject.mockResolvedValue({
  html_url: 'https://github.com/web/site'
});

// @ts-ignore
jest.mock('../git.ts', () => (...args) => {
  constructor(...args);
  return {
    options: { owner: 'test', repo: 'test', version: '1.0.0' },
    getGitLog,
    publish,
    getLabels,
    getLatestRelease,
    getPullRequest,
    getSha,
    createStatus,
    createComment,
    getProject,
    changedPackages,
    getCommitsForPR,
    getUserByUsername,
    getUserByEmail,
    getProjectLabels,
    createLabel,
    getPullRequests
  };
});

const execSpy = jest.fn();
// @ts-ignore
jest.mock('../utils/exec-promise.ts', () => (...args) => execSpy(...args));

const existsSync = jest.fn();
const writeSpy = jest.fn();

let readResult = '{}';

const options = { owner: 'Adam Dierkens', repo: 'test', token: 'MY_TOKENss' };

jest.mock('fs', () => ({
  // @ts-ignore
  existsSync: (...args) => existsSync(...args),
  // @ts-ignore
  readFile: (a, b, cb) => {
    cb(null, readResult);
  },
  // @ts-ignore
  ReadStream: () => undefined,
  // @ts-ignore
  WriteStream: () => undefined,
  // @ts-ignore
  closeSync: () => undefined,
  // @ts-ignore
  writeFile: (file, data, cb) => {
    cb(null, writeSpy(file, data));
  }
}));

const slackSpy = jest.fn();

jest.mock('../utils/slack.ts', () => () => slackSpy());
jest.mock('../plugins/npm/package-config.ts', () => async () => ({}));
jest.mock('../utils/github-token', () => async () => ({}));

describe('GitHubRelease', () => {
  beforeEach(() => {
    getGitLog.mockClear();
    publish.mockClear();
    getLabels.mockClear();
    writeSpy.mockClear();
    execSpy.mockClear();
    execSpy.mockClear();
    createLabel.mockClear();
  });

  test('should use options owner, repo, and token', async () => {
    const gh = new GitHubRelease({
      owner: 'Andrew',
      repo: 'test',
      token: 'MY_TOKENss'
    });

    await gh.getCommits('12345');

    expect(constructor.mock.calls[0][0].owner).toBe('Andrew');
    expect(constructor.mock.calls[0][0].repo).toBe('test');
  });

  test('should throw without options owner, repo, and token', async () => {
    expect(() => new GitHubRelease({})).toThrow();
  });

  describe('getCommits', async () => {
    test('should default to HEAD', async () => {
      const gh = new GitHubRelease(options);
      await gh.getCommits('12345');
      expect(getGitLog).toHaveBeenCalled();
    });

    test('should use configured HEAD', async () => {
      const gh = new GitHubRelease({
        owner: 'Adam Dierkens',
        repo: 'test',
        token: 'test'
      });
      await gh.getCommits('12345', '1234');
      expect(getGitLog).toHaveBeenCalled();
    });

    test('should not resolve authors with no PR commits', async () => {
      const commits = [
        makeCommitFromMsg('First'),
        makeCommitFromMsg('Second'),
        makeCommitFromMsg('Third')
      ];

      getGitLog.mockReturnValueOnce(commits);
      const gh = new GitHubRelease({
        owner: 'Adam Dierkens',
        repo: 'test',
        token: 'test'
      });
      await gh.getCommits('12345', '1234');
      expect(getUserByUsername).not.toHaveBeenCalled();
    });

    test('should resolve authors with PR commits', async () => {
      const commits = normalizeCommits([
        makeCommitFromMsg('First'),
        makeCommitFromMsg('Second (#123)', {
          name: 'Andrew Lisowski',
          email: 'lisowski54@gmail.com'
        }),
        makeCommitFromMsg('Third')
      ]);

      getGitLog.mockReturnValueOnce(commits);
      getCommitsForPR.mockReturnValueOnce([
        {
          author: {
            login: 'andrew'
          }
        }
      ]);
      getUserByUsername.mockReturnValueOnce({
        login: 'andrew',
        name: 'Andrew Lisowski'
      });
      getUserByEmail.mockReturnValueOnce({
        login: 'adam',
        name: 'Adam Dierkens'
      });
      getUserByEmail.mockReturnValueOnce({
        login: 'adam',
        name: 'Adam Dierkens'
      });

      const gh = new GitHubRelease({
        owner: 'Adam Dierkens',
        repo: 'test',
        token: 'test'
      });
      const modifiedCommits = await gh.getCommits('12345', '1234');
      expect(getUserByUsername).toHaveBeenCalled();
      expect(modifiedCommits).toMatchSnapshot();
    });
  });

  test('publish', async () => {
    const gh = new GitHubRelease(options);
    await gh.publish('release notes', '1.2.3');
    expect(publish).toHaveBeenCalled();
  });

  test('getLabels', async () => {
    const gh = new GitHubRelease(options);
    await gh.getLabels(123);
    expect(getLabels).toHaveBeenCalled();
  });

  test('getLatestRelease', async () => {
    const gh = new GitHubRelease(options);
    await gh.getLatestRelease();
    expect(getLatestRelease).toHaveBeenCalled();
  });

  test('getLatestRelease', async () => {
    const gh = new GitHubRelease(options);
    await gh.getPullRequest(22);
    expect(getLatestRelease).toHaveBeenCalled();
  });

  test('getSha', async () => {
    const gh = new GitHubRelease(options);
    await gh.getSha();
    expect(getLatestRelease).toHaveBeenCalled();
  });

  test('createStatus', async () => {
    const gh = new GitHubRelease(options);
    await gh.createStatus({
      state: 'pending',
      sha: '',
      context: 'foo',
      target_url: 'google.com',
      description: 'testing'
    });
    expect(getLatestRelease).toHaveBeenCalled();
  });

  test('createComment', async () => {
    const gh = new GitHubRelease(options);
    await gh.createComment('Some long message', 22);
    expect(createComment).toHaveBeenCalled();
  });

  test('getPullRequests', async () => {
    const gh = new GitHubRelease(options);
    await gh.getPullRequests({ state: 'closed' });
    expect(getPullRequests).toHaveBeenCalled();
  });

  describe('addToChangelog', async () => {
    test("creates new changelog if one didn't exist", async () => {
      const gh = new GitHubRelease(options);
      await gh.addToChangelog(
        '# My new Notes',
        'klajsdlfk4lj51l43k5hj234l',
        'v0.0.0'
      );

      expect(writeSpy.mock.calls[0][1].includes(`# My new Notes`)).toBe(true);
    });

    test("creates new changelog if one didn't exist", async () => {
      const gh = new GitHubRelease(options);
      await gh.addToChangelog('# My new Notes', 'v1.0.0', 'v1.0.0');

      expect(writeSpy.mock.calls[0][1].includes(`v1.0.1`)).toBe(true);
    });

    test('creates changelog with v in versions', async () => {
      const gh = new GitHubRelease(options, {
        noVersionPrefix: true,
        skipReleaseLabels: ['skip-release']
      });
      await gh.addToChangelog('# My new Notes', '1.0.0', '1.0.0');

      expect(writeSpy.mock.calls[0][1].includes(`1.0.1`)).toBe(true);
    });

    test('prepends to old changelog', async () => {
      const gh = new GitHubRelease(options);

      existsSync.mockReturnValueOnce(true);
      readResult = '# My old Notes';

      await gh.addToChangelog(
        '# My new Notes',
        'asdfasdlkfjlkj435l2j',
        'v0.0.0'
      );
      expect(writeSpy.mock.calls[0][1].includes(readResult)).toBe(true);
    });

    test('should be able to configure message', async () => {
      const gh = new GitHubRelease(options);
      const message = 'pony foo';

      existsSync.mockReturnValueOnce(true);
      readResult = '# My old Notes';

      await gh.addToChangelog(
        '# My new Notes',
        'asdklfhlkh24387513',
        'v0.0.0',
        message
      );
      expect(execSpy.mock.calls[1][1].includes(`"${message}"`)).toBe(true);
    });
  });

  describe('postToSlack', () => {
    test('throws without slack url', async () => {
      const gh = new GitHubRelease(options, {
        skipReleaseLabels: []
      });
      expect(gh.postToSlack('# My Notes', 'v1.0.0')).rejects.toBeTruthy();
    });

    test('successful', async () => {
      const gh = new GitHubRelease(options, {
        slack: 'https://custom-slack-url',
        skipReleaseLabels: []
      });
      await gh.postToSlack('# My Notes', 'v1.0.0');
      expect(slackSpy).toHaveBeenCalled();
    });
  });

  describe('generateReleaseNotes', async () => {
    test('should default to HEAD', async () => {
      const gh = new GitHubRelease(options);
      expect(await gh.generateReleaseNotes('1234')).toBe('');
    });

    test('should use configured HEAD', async () => {
      const gh = new GitHubRelease(options);
      expect(await gh.generateReleaseNotes('1234', '123')).toBe('');
    });

    test('should allow user to configure section headings', async () => {
      const gh = new GitHubRelease(options);

      const commits = [
        makeCommitFromMsg('First (#1234)'),
        makeCommitFromMsg('Second (#1235)'),
        makeCommitFromMsg('Third (#1236)'),
        makeCommitFromMsg('Fourth (#1237)')
      ];

      getGitLog.mockReturnValueOnce(commits);
      getLabels.mockReturnValueOnce(['major']);
      getLabels.mockReturnValueOnce(['minor']);
      getLabels.mockReturnValueOnce(['documentation', 'internal']);
      getLabels.mockReturnValueOnce(['patch']);

      expect(await gh.generateReleaseNotes('1234', '123')).toMatchSnapshot();
    });
  });

  describe('getSemverBump', () => {
    test('default to patch', async () => {
      const gh = new GitHubRelease(options);
      const commits = [
        makeCommitFromMsg('First'),
        makeCommitFromMsg('Second'),
        makeCommitFromMsg('Third')
      ];

      getGitLog.mockReturnValueOnce(commits);

      expect(await gh.getSemverBump('1234')).toBe(SEMVER.patch);
    });

    test('should use higher version', async () => {
      const gh = new GitHubRelease(options);
      const commits = [
        makeCommitFromMsg('First (#1234)'),
        makeCommitFromMsg('Second'),
        makeCommitFromMsg('Third')
      ];

      getGitLog.mockReturnValueOnce(commits);
      getLabels.mockReturnValueOnce(['minor']);

      expect(await gh.getSemverBump('1234', '123')).toBe(SEMVER.minor);
    });

    test('should not publish a release', async () => {
      const gh = new GitHubRelease(options);
      const commits = [
        makeCommitFromMsg('First (#1234)'),
        makeCommitFromMsg('Second (#1235)'),
        makeCommitFromMsg('Third (#1236)')
      ];

      getGitLog.mockReturnValueOnce(commits);
      getLabels.mockReturnValueOnce(['skip-release', 'patch']);
      getLabels.mockReturnValueOnce(['skip-release', 'patch']);
      getLabels.mockReturnValueOnce(['skip-release', 'minor']);

      expect(await gh.getSemverBump('1234', '123')).toBe('');
    });

    test('should publish a release', async () => {
      const gh = new GitHubRelease(options);
      const commits = [
        makeCommitFromMsg('First (#1234)'),
        makeCommitFromMsg('Second (#1235)'),
        makeCommitFromMsg('Third (#1236)')
      ];

      getGitLog.mockReturnValueOnce(commits);
      getLabels.mockReturnValueOnce(['patch']);
      getLabels.mockReturnValueOnce(['skip-release', 'patch']);
      getLabels.mockReturnValueOnce(['skip-release', 'minor']);

      expect(await gh.getSemverBump('1234', '123')).toBe(SEMVER.minor);
    });

    test('should default to publish a prepatch', async () => {
      const gh = new GitHubRelease(options);
      const commits = [
        makeCommitFromMsg('First (#1234)'),
        makeCommitFromMsg('Second (#1235)'),
        makeCommitFromMsg('Third (#1236)')
      ];

      getGitLog.mockReturnValueOnce(commits);
      getLabels.mockReturnValueOnce(['skip-release']);
      getLabels.mockReturnValueOnce([]);
      getLabels.mockReturnValueOnce([]);

      expect(await gh.getSemverBump('1234', '123')).toBe('');
    });

    test('should not publish a release in onlyPublishWithReleaseLabel without label', async () => {
      const gh = new GitHubRelease(options, {
        onlyPublishWithReleaseLabel: true,
        skipReleaseLabels: []
      });
      const commits = [
        makeCommitFromMsg('First (#1234)'),
        makeCommitFromMsg('Second (#1235)'),
        makeCommitFromMsg('Third (#1236)')
      ];

      getGitLog.mockReturnValueOnce(commits);
      getLabels.mockReturnValueOnce(['patch']);
      getLabels.mockReturnValueOnce(['major']);
      getLabels.mockReturnValueOnce(['patch']);

      expect(await gh.getSemverBump('1234', '123')).toBe('');
    });

    test('should publish a release in onlyPublishWithReleaseLabel with label', async () => {
      const gh = new GitHubRelease(options, {
        onlyPublishWithReleaseLabel: true,
        skipReleaseLabels: []
      });
      const commits = [
        makeCommitFromMsg('First (#1234)'),
        makeCommitFromMsg('Second (#1235)'),
        makeCommitFromMsg('Third (#1236)')
      ];

      getGitLog.mockReturnValueOnce(commits);
      getLabels.mockReturnValueOnce(['release', 'patch']);
      getLabels.mockReturnValueOnce(['patch']);
      getLabels.mockReturnValueOnce(['minor']);

      expect(await gh.getSemverBump('1234', '123')).toBe(SEMVER.minor);
    });

    test('should be able to configure labels', async () => {
      const versionLabels = new Map();
      versionLabels.set(SEMVER.major, 'Version: Major');
      versionLabels.set(SEMVER.minor, 'Version: Minor');
      versionLabels.set(SEMVER.patch, 'Version: Patch');
      versionLabels.set('release', 'Deploy');

      const gh = new GitHubRelease(options, {
        onlyPublishWithReleaseLabel: true,
        skipReleaseLabels: [],
        versionLabels
      });
      const commits = [
        makeCommitFromMsg('First (#1234)'),
        makeCommitFromMsg('Second (#1235)'),
        makeCommitFromMsg('Third (#1236)')
      ];

      // Test default labels do nothing
      getGitLog.mockReturnValueOnce(commits);
      getLabels.mockReturnValueOnce(['Version: Major']);
      getLabels.mockReturnValueOnce(['Version: Patch']);
      getLabels.mockReturnValueOnce(['Version: Minor', 'release']);

      expect(await gh.getSemverBump('1234', '123')).toBe('');

      getGitLog.mockReturnValueOnce(commits);
      getLabels.mockReturnValueOnce(['Version: Minor', 'Deploy']);
      getLabels.mockReturnValueOnce(['Version: Major']);
      getLabels.mockReturnValueOnce(['Version: Patch']);

      expect(await gh.getSemverBump('1234', '123')).toBe(SEMVER.major);
    });
  });

  describe('addLabelsToProject', () => {
    test('should add labels', async () => {
      const gh = new GitHubRelease(options);
      const labels = new Labels();
      labels.major = '1';
      labels.minor = '2';
      labels.patch = '3';

      await gh.addLabelsToProject(labels);

      expect(createLabel).toHaveBeenCalledWith(SEMVER.major, '1');
      expect(createLabel).toHaveBeenCalledWith(SEMVER.minor, '2');
      expect(createLabel).toHaveBeenCalledWith(SEMVER.patch, '3');
    });

    test('should log that it has created the labels', async () => {
      const mockLogger = dummyLog();
      mockLogger.log.log = jest.fn();

      const gh = new GitHubRelease(
        options,
        {
          skipReleaseLabels: []
        },
        mockLogger
      );
      const labels = new Labels();
      labels.patch = '3';

      await gh.addLabelsToProject(labels);

      expect(mockLogger.log.log).toHaveBeenCalledWith('Created labels: patch');
      expect(mockLogger.log.log).toHaveBeenCalledWith(
        '\nYou can see these, and more at https://github.com/web/site/labels'
      );
    });

    test('should not add old labels', async () => {
      const gh = new GitHubRelease(options);
      const labels = new Labels();
      labels.major = '1';
      labels.minor = '2';

      getProjectLabels.mockReturnValueOnce(['1']);
      await gh.addLabelsToProject(labels);

      expect(createLabel).not.toHaveBeenCalledWith(SEMVER.major, '1');
      expect(createLabel).toHaveBeenCalledWith(SEMVER.minor, '2');
    });

    test('should add release label in onlyPublishWithReleaseLabel mode', async () => {
      let gh = new GitHubRelease(options, {
        skipReleaseLabels: []
      });
      const labels = new Labels();
      labels.release = 'deploy';

      await gh.addLabelsToProject(labels);
      expect(createLabel).not.toHaveBeenCalledWith('release', 'deploy');

      gh = new GitHubRelease(options, {
        onlyPublishWithReleaseLabel: true,
        skipReleaseLabels: []
      });
      await gh.addLabelsToProject(labels);
      expect(createLabel).toHaveBeenCalledWith('release', 'deploy');
    });

    test('should add skip-release label not in onlyPublishWithReleaseLabel mode', async () => {
      let gh = new GitHubRelease(options, {
        onlyPublishWithReleaseLabel: true,
        skipReleaseLabels: []
      });
      const labels = new Labels();
      labels['skip-release'] = 'no!';

      await gh.addLabelsToProject(labels);
      expect(createLabel).not.toHaveBeenCalledWith('skip-release', 'no!');

      gh = new GitHubRelease(options, {
        skipReleaseLabels: []
      });
      await gh.addLabelsToProject(labels);
      expect(createLabel).toHaveBeenCalledWith('skip-release', 'no!');
    });
  });
});
