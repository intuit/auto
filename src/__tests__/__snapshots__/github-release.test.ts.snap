// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitHubRelease generateReleaseNotes should allow user to configure section headings 1`] = `
"#### 💥  Breaking Change

- First [#1234](https://github.com/web/site/pull/1234) (adam@dierkens.com)

#### 🚀  Enhancement

- Second [#1235](https://github.com/web/site/pull/1235) (adam@dierkens.com)

#### 🐛  Bug Fix

- Fourth [#1237](https://github.com/web/site/pull/1237) (adam@dierkens.com)

#### 🏠  Internal

- Third [#1236](https://github.com/web/site/pull/1236) (adam@dierkens.com)

#### Authors: 1

- Adam Dierkens (adam@dierkens.com)"
`;

exports[`GitHubRelease generateReleaseNotes should include PR-less commits 1`] = `
"#### 🚀  Enhancement

- First Feature [#1235](https://github.com/web/site/pull/1235) (adam@dierkens.com)

#### ⚠️  Pushed to master

- I should be included  (adam@dierkens.com)

#### Authors: 2

- Adam Dierkens (adam@dierkens.com)
- adam@dierkens.com"
`;

exports[`GitHubRelease generateReleaseNotes should match rebased commits to PRs 1`] = `
"#### 🚀  Enhancement

- Feature [#124](https://github.com/web/site/pull/124) (adam@dierkens.com)
- I was rebased [#123](https://github.com/web/site/pull/123) (adam@dierkens.com)

#### ⚠️  Pushed to master

- I am a commit to master  (adam@dierkens.com)

#### Authors: 2

- Adam Dierkens (adam@dierkens.com)
- adam@dierkens.com"
`;

exports[`GitHubRelease getCommits should ignore rebased commits if no last release 1`] = `
Array [
  Object {
    "authorEmail": "adam@dierkens.com",
    "authorName": "Adam Dierkens",
    "authors": Array [
      Object {
        "email": "adam@dierkens.com",
        "name": "Adam Dierkens",
      },
    ],
    "hash": "foo",
    "labels": Array [],
    "packages": undefined,
    "pullRequest": Object {
      "number": 123,
    },
    "subject": "Second",
  },
]
`;

exports[`GitHubRelease getCommits should match rebased commits to PRs 1`] = `
Array [
  Object {
    "authorEmail": "adam@dierkens.com",
    "authorName": "Adam Dierkens",
    "authors": Array [
      Object {
        "email": "adam@dierkens.com",
        "name": "Adam Dierkens",
      },
    ],
    "hash": "foo",
    "labels": Array [],
    "packages": undefined,
    "pullRequest": Object {
      "number": 124,
    },
    "subject": "Feature",
  },
  Object {
    "authorEmail": "adam@dierkens.com",
    "authorName": "Adam Dierkens",
    "authors": Array [
      Object {
        "email": "adam@dierkens.com",
        "name": "Adam Dierkens",
      },
    ],
    "hash": "1a2b",
    "labels": Array [
      "skip-release",
      "minor",
    ],
    "packages": undefined,
    "pullRequest": Object {
      "number": 123,
    },
    "subject": "I was rebased",
  },
]
`;

exports[`GitHubRelease getCommits should match rebased commits to PRs with first commit 1`] = `
Array [
  Object {
    "authorEmail": "adam@dierkens.com",
    "authorName": "Adam Dierkens",
    "authors": Array [
      Object {
        "email": "adam@dierkens.com",
        "name": "Adam Dierkens",
      },
    ],
    "hash": "foo",
    "labels": Array [],
    "packages": undefined,
    "pullRequest": Object {
      "number": 124,
    },
    "subject": "Feature",
  },
  Object {
    "authorEmail": "adam@dierkens.com",
    "authorName": "Adam Dierkens",
    "authors": Array [
      Object {
        "email": "adam@dierkens.com",
        "name": "Adam Dierkens",
      },
    ],
    "hash": "1a2b",
    "labels": Array [
      "skip-release",
      "minor",
    ],
    "packages": undefined,
    "pullRequest": Object {
      "number": 123,
    },
    "subject": "I was rebased",
  },
]
`;

exports[`GitHubRelease getCommits should resolve authors with PR commits 1`] = `
Array [
  Object {
    "authorEmail": "adam@dierkens.com",
    "authorName": "Adam Dierkens",
    "authors": Array [
      Object {
        "login": "adam",
        "name": "Adam Dierkens",
        "username": "adam",
      },
    ],
    "hash": "foo",
    "labels": Array [],
    "packages": undefined,
    "subject": "First",
  },
  Object {
    "authorEmail": "lisowski54@gmail.com",
    "authorName": "Andrew Lisowski",
    "authors": Array [
      Object {
        "login": "andrew",
        "name": "Andrew Lisowski",
        "username": "andrew",
      },
    ],
    "hash": "foo",
    "labels": Array [],
    "packages": undefined,
    "pullRequest": Object {
      "number": 123,
    },
    "subject": "Second",
  },
  Object {
    "authorEmail": "adam@dierkens.com",
    "authorName": "Adam Dierkens",
    "authors": Array [
      Object {
        "login": "adam",
        "name": "Adam Dierkens",
        "username": "adam",
      },
    ],
    "hash": "foo",
    "labels": Array [],
    "packages": undefined,
    "subject": "Third",
  },
]
`;
